File: kotlin-node/karakum/inheritanceModifiers/modifyClassInheritance.js
import ts from "typescript"

export default (node) => {
    const sourceFileName = node.getSourceFile()?.fileName ?? "generated.d.ts"

    if (
        ts.isClassDeclaration(node)
        && node.name
        && (
            (
                sourceFileName.endsWith("crypto.d.ts")
                && (
                    node.name.text === "Cipher"
                    || node.name.text === "Decipher"
                )
            )
            || (
                sourceFileName.endsWith("readline.d.ts")
                && (
                    node.name.text === "Interface"
                )
            )
            || (
                sourceFileName.endsWith("child_process.d.ts")
                && (
                    node.name.text === "ChildProcess"
                )
            )
            || (
                sourceFileName.endsWith("http.d.ts")
                && (
                    node.name.text === "IncomingMessage"
                    || node.name.text === "OutgoingMessage"
                    || node.name.text === "ServerResponse"
                    || node.name.text === "Agent"
                    || node.name.text === "Server"
                )
            )
            || (
                sourceFileName.endsWith("http2.d.ts")
                && (
                    node.name.text === "Http2ServerRequest"
                    || node.name.text === "Http2ServerResponse"
                )
            )
            || (
                sourceFileName.endsWith("net.d.ts")
                && (
                    node.name.text === "Server"
                    || node.name.text === "Socket"
                )
            )
            || (
                sourceFileName.endsWith("stream.d.ts")
                && (
                    node.name.text === "ReadableBase"
                    || node.name.text === "WritableBase"
                    || node.name.text === "Readable"
                    || node.name.text === "Writable"
                    || node.name.text === "Duplex"
                    || node.name.text === "Transform"
                )
            )
            || (
                sourceFileName.endsWith("tty.d.ts")
                && (
                    node.name.text === "ReadStream"
                    || node.name.text === "WriteStream"
                )
            )
            || (
                sourceFileName.endsWith("async_hooks.d.ts")
                && (
                    node.name.text === "AsyncResource"
                )
            )
            || (
                sourceFileName.endsWith("perf_hooks.d.ts")
                && (
                    node.name.text === "PerformanceEntry"
                )
            )
            || (
                sourceFileName.endsWith("vm.d.ts")
                && (
                    node.name.text === "Module"
                )
            )
            || (
                sourceFileName.endsWith("tls.d.ts")
                && (
                    node.name.text === "Server"
                )
            )
            || (
                sourceFileName.endsWith("v8.d.ts")
                && (
                    node.name.text === "Serializer"
                    || node.name.text === "Deserializer"
                )
            )
        )
    ) {
        return "open"
    }

    return null
}

File: kotlin-node/karakum/src/jsMain/kotlin/node/karakum/inheritanceModifiers/modifyClassInheritance.kt
package node.karakum.inheritanceModifiers

import io.github.sgrishchenko.karakum.extension.InheritanceModifierContext
import io.github.sgrishchenko.karakum.util.getSourceFileOrNull
import node.karakum.util.nullable
import typescript.Node
import typescript.isClassDeclaration

fun modifyClassInheritance(node: Node, context: InheritanceModifierContext) = nullable {
    val sourceFileName = ensureNotNull(node.getSourceFileOrNull()).fileName

    ensure(isClassDeclaration(node))

    nullable {
        ensure(sourceFileName.endsWith("crypto.d.ts"))
        ensure(
            node.name?.text == "Cipheriv"
                    || node.name?.text == "Decipheriv"
        )

        "open"
    } ?: nullable {
        ensure(sourceFileName.endsWith("readline.d.ts"))
        ensure(node.name?.text == "Interface")

        "open"
    } ?: nullable {
        ensure(sourceFileName.endsWith("child_process.d.ts"))
        ensure(node.name?.text == "ChildProcess")

        "open"
    } ?: nullable {
        ensure(sourceFileName.endsWith("http.d.ts"))
        ensure(
            node.name?.text == "IncomingMessage"
                    || node.name?.text == "OutgoingMessage"
                    || node.name?.text == "ServerResponse"
                    || node.name?.text == "Agent"
                    || node.name?.text == "Server"
        )

        "open"
    } ?: nullable {
        ensure(sourceFileName.endsWith("http2.d.ts"))
        ensure(
            node.name?.text == "Http2ServerRequest"
                    || node.name?.text == "Http2ServerResponse"
        )

        "open"
    } ?: nullable {
        ensure(sourceFileName.endsWith("net.d.ts"))
        ensure(
            node.name?.text == "Server"
                    || node.name?.text == "Socket"
        )

        "open"
    } ?: nullable {
        ensure(sourceFileName.endsWith("stream.d.ts"))
        ensure(
            node.name?.text == "Stream"
                    || node.name?.text == "Readable"
                    || node.name?.text == "Writable"
                    || node.name?.text == "Duplex"
                    || node.name?.text == "Transform"
        )

        "open"
    } ?: nullable {
        ensure(sourceFileName.endsWith("tty.d.ts"))
        ensure(
            node.name?.text == "ReadStream"
                    || node.name?.text == "WriteStream"
        )

        "open"
    } ?: nullable {
        ensure(sourceFileName.endsWith("async_hooks.d.ts"))
        ensure(node.name?.text == "AsyncResource")

        "open"
    } ?: nullable {
        ensure(sourceFileName.endsWith("perf_hooks.d.ts"))
        ensure(node.name?.text == "PerformanceEntry")

        "open"
    } ?: nullable {
        ensure(sourceFileName.endsWith("vm.d.ts"))
        ensure(node.name?.text == "Module")

        "open"
    } ?: nullable {
        ensure(sourceFileName.endsWith("tls.d.ts"))
        ensure(node.name?.text == "Server")

        "open"
    } ?: nullable {
        ensure(sourceFileName.endsWith("v8.d.ts"))
        ensure(
            node.name?.text == "Serializer"
                    || node.name?.text == "Deserializer"
        )

        "open"
    }
}
